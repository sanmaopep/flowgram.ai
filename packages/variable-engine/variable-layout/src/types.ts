import { Scope } from '@flowgram.ai/variable-core';
import { FlowNodeEntity } from '@flowgram.ai/document';

export enum FlowNodeScopeTypeEnum {
  public = 'public',
  private = 'private',
}

export interface FlowNodeScopeMeta {
  node?: FlowNodeEntity;
  type?: FlowNodeScopeTypeEnum;
}

export interface ScopeVirtualNode {
  id: string;
  flowNodeType: 'virtualNode';
}

export type ScopeChainNode = FlowNodeEntity | ScopeVirtualNode;

// 节点内部的作用域
export interface FlowNodeScope extends Scope<FlowNodeScopeMeta> {}
