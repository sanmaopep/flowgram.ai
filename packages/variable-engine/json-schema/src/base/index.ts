/**
 * Copyright (c) 2025 Bytedance Ltd. and/or its affiliates
 * SPDX-License-Identifier: MIT
 */

export { BaseTypeManager } from './base-type-manager';
export { BaseTypeRegistry, TypeRegistryCreator } from './types';
