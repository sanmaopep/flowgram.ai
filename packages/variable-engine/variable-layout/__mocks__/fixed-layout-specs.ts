import { FlowDocumentJSON } from '@flowgram.ai/document';

export const fixLayout1: FlowDocumentJSON = {
  nodes: [
    {
      type: 'start',
      meta: {
        isStart: true,
      },
      id: 'start',
    },
    {
      type: 'getRecords',
      id: 'getRecords_07e97c55832',
    },
    {
      type: 'loop',
      id: 'forEach_260a8f85ff2',
      blocks: [
        {
          type: 'createRecord',
          id: 'createRecord_8f85ff2c11d',
        },
        {
          type: 'dynamicSplit',
          id: 'exclusiveSplit_ff2c11d0fb4',
          blocks: [
            {
              id: 'branch_f2c11d0fb42',
            },
            {
              id: 'branch_2c11d0fb42c',
            },
          ],
        },
      ],
    },
    {
      type: 'dynamicSplit',
      id: 'exclusiveSplit_88dbf2c60ae',
      meta: {
        defaultExpanded: true,
      },
      blocks: [
        {
          id: 'branch_8dbf2c60aee',
          blocks: [
            {
              type: 'dynamicSplit',
              id: 'exclusiveSplit_a59afaadc9a',
              blocks: [
                {
                  id: 'branch_59afaadc9ac',
                },
                {
                  id: 'branch_9afaadc9acd',
                  blocks: [
                    {
                      type: 'deleteRecords',
                      id: 'deleteRecords_c32807e97c5',
                    },
                  ]
                },
              ],
            },
          ]
        },
        {
          id: 'branch_dbf2c60aee4',
          blocks: [
            {
              type: 'updateRecords',
              id: 'updateRecords_7ed2a172c32',
            },
          ]
        },
      ],
    },
    {
      type: 'end',
      id: 'end',
    },
  ],
};
