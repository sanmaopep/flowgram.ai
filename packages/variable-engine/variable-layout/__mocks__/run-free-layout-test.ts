import { describe, expect, test } from 'vitest';
import { VariableEngine } from '@flowgram.ai/variable-core';
import { ASTKind } from '@flowgram.ai/variable-core';
import {
  FlowNodeVariableData
} from '../src';
import { TestConfig, getContainer } from './container';
import { WorkflowDocument, WorkflowJSON } from '@flowgram.ai/free-layout-core';

export const runFreeLayoutTest = (testName: string, spec: WorkflowJSON, config?: TestConfig) => {
  describe(testName, async () => {
    const container = getContainer('free', config);
    const flowDocument = container.get(WorkflowDocument);
    const variableEngine = container.get(VariableEngine);

    variableEngine.onScopeChange(action => {
      const { type, scope } = action;

      // 作用域默认创建一个默认变量
      if (type === 'add') {
        scope.ast.set('/', {
          kind: ASTKind.VariableDeclaration,
          type: ASTKind.String,
          key: String(scope.id),
        });
      }
    });

    // 创建一个全局作用域
    variableEngine.createScope('testScope');

    const traverseVariableDatas = () =>
      flowDocument
        .getAllNodes()
        // TODO 包含 $ 的节点不注册 variableData
        .filter(_node => !_node.id.startsWith('$'))
        .map(_node => _node.getData(FlowNodeVariableData))
        .filter(Boolean);

    // 初始化所有节点的私有作用域
    const initAllNodePrivate = () => {
      traverseVariableDatas().forEach(_data => {
        _data.initPrivate();
      });
    };

    // 初始化所有节点的可用变量
    const printAllNodeAvailableMapping = (_scopeType: 'public' | 'private' = 'public') =>
      traverseVariableDatas().reduce((acm, _data) => {
        const scope = _data[_scopeType]!;
        acm.set(String(scope.id), scope.available.variableKeys);

        return acm;
      }, new Map<string, string[]>());

    // 初始化所有节点的覆盖作用域
    const printAllCovers = (_scopeType: 'public' | 'private' = 'public') =>
      traverseVariableDatas().reduce((acm, _data) => {
        const scope = _data[_scopeType]!;
        acm.set(
          String(scope.id),
          scope.coverScopes.map(_scope => String(_scope.id)),
        );

        return acm;
      }, new Map<string, string[]>());

    await flowDocument.fromJSON(spec);

    test('test get Deps', () => {
      expect(printAllNodeAvailableMapping()).toMatchSnapshot();
    });

    test('test get Covers', () => {
      expect(printAllCovers()).toMatchSnapshot();
    });

    test('test get Deps After Init Private', () => {
      initAllNodePrivate();
      expect(printAllNodeAvailableMapping()).toMatchSnapshot();
    });

    test('test get private scope Deps', () => {
      expect(printAllNodeAvailableMapping('private')).toMatchSnapshot();
    });

    test('test get Covers After Init Private', () => {
      expect(printAllCovers()).toMatchSnapshot();
    });

    test('test get private scope Covers', () => {
      expect(printAllCovers('private')).toMatchSnapshot();
    });

    test('test sort', () => {
      expect(variableEngine.getAllScopes({ sort: true }).map(_scope => _scope.id)).toMatchSnapshot();
    });

    config?.runExtraTest?.(container);
  });


}
