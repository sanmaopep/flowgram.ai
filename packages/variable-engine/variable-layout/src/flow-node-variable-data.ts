import { BaseVariableField, VariableEngine } from '@flowgram.ai/variable-core';
import { type ASTNode, ASTNodeJSON } from '@flowgram.ai/variable-core';
import { FlowNodeEntity } from '@flowgram.ai/document';
import { EntityData } from '@flowgram.ai/core';

import { FlowNodeScope, FlowNodeScopeMeta, FlowNodeScopeTypeEnum } from './types';

interface Options {
  variableEngine: VariableEngine;
}

export class FlowNodeVariableData extends EntityData {
  static type: string = 'FlowNodeVariableData';

  declare entity: FlowNodeEntity;

  readonly variableEngine: VariableEngine;

  /**
   * Private variables can be accessed by public ones, but not the other way around.
   */
  protected _private?: FlowNodeScope;

  protected _public: FlowNodeScope;

  get private() {
    return this._private;
  }

  get public() {
    return this._public;
  }

  /**
   * Sets a variable in the public AST (Abstract Syntax Tree) with the given key and JSON value.
   *
   * @param key - The key under which the variable will be stored.
   * @param json - The JSON value to store.
   * @returns The updated AST node.
   */
  public setVar(key: string, json: ASTNodeJSON): ASTNode;

  /**
   * Sets a variable in the public AST (Abstract Syntax Tree) with the default key 'outputs'.
   *
   * @param json - The JSON value to store.
   * @returns The updated AST node.
   */
  public setVar(json: ASTNodeJSON): ASTNode;

  public setVar(arg1: string | ASTNodeJSON, arg2?: ASTNodeJSON): ASTNode {
    if (typeof arg1 === 'string' && arg2 !== undefined) {
      return this.public.ast.set(arg1, arg2);
    }

    if (typeof arg1 === 'object' && arg2 === undefined) {
      return this.public.ast.set('outputs', arg1);
    }

    throw new Error('Invalid arguments');
  }

  /**
   * Retrieves a variable from the public AST (Abstract Syntax Tree) by key.
   *
   * @param key - The key of the variable to retrieve. Defaults to 'outputs'.
   * @returns The value of the variable, or undefined if not found.
   */
  public getVar(key: string = 'outputs') {
    return this.public.ast.get(key);
  }

  /**
   * Clears a variable from the public AST (Abstract Syntax Tree) by key.
   *
   * @param key - The key of the variable to clear. Defaults to 'outputs'.
   * @returns The updated AST node.
   */
  public clearVar(key: string = 'outputs') {
    return this.public.ast.remove(key);
  }

  /**
   * Sets a variable in the private AST (Abstract Syntax Tree) with the given key and JSON value.
   *
   * @param key - The key under which the variable will be stored.
   * @param json - The JSON value to store.
   * @returns The updated AST node.
   */
  public setPrivateVar(key: string, json: ASTNodeJSON): ASTNode;

  /**
   * Sets a variable in the private AST (Abstract Syntax Tree) with the default key 'outputs'.
   *
   * @param json - The JSON value to store.
   * @returns The updated AST node.
   */
  public setPrivateVar(json: ASTNodeJSON): ASTNode;

  public setPrivateVar(arg1: string | ASTNodeJSON, arg2?: ASTNodeJSON): ASTNode {
    if (typeof arg1 === 'string' && arg2 !== undefined) {
      return this.initPrivate().ast.set(arg1, arg2);
    }

    if (typeof arg1 === 'object' && arg2 === undefined) {
      return this.initPrivate().ast.set('outputs', arg1);
    }

    throw new Error('Invalid arguments');
  }

  /**
   * Retrieves a variable from the private AST (Abstract Syntax Tree) by key.
   *
   * @param key - The key of the variable to retrieve. Defaults to 'outputs'.
   * @returns The value of the variable, or undefined if not found.
   */
  public getPrivateVar(key: string = 'outputs') {
    return this.private?.ast.get(key);
  }

  /**
   * Clears a variable from the private AST (Abstract Syntax Tree) by key.
   *
   * @param key - The key of the variable to clear. Defaults to 'outputs'.
   * @returns The updated AST node.
   */
  public clearPrivateVar(key: string = 'outputs') {
    return this.private?.ast.remove(key);
  }

  get allScopes(): FlowNodeScope[] {
    const res = [];

    if (this._public) {
      res.push(this._public);
    }
    if (this._private) {
      res.push(this._private);
    }

    return res;
  }

  getDefaultData() {
    return {};
  }

  constructor(entity: FlowNodeEntity, readonly opts: Options) {
    super(entity);

    const { variableEngine } = opts || {};
    this.variableEngine = variableEngine;
    this._public = this.variableEngine.createScope(this.entity.id, {
      node: this.entity,
      type: FlowNodeScopeTypeEnum.public,
    } as FlowNodeScopeMeta);
    this.toDispose.push(this._public);
  }

  initPrivate(): FlowNodeScope {
    if (!this._private) {
      this._private = this.variableEngine.createScope(`${this.entity.id}_private`, {
        node: this.entity,
        type: FlowNodeScopeTypeEnum.private,
      } as FlowNodeScopeMeta);
      // 1. Notify the covering scopes of private to update dependencies
      this._private.coverScopes.forEach((_scope) => {
        _scope.refreshDeps();
      });
      // 2. Notify the dependent scopes of private to update their covers
      this._private.depScopes.forEach((_scope) => {
        _scope.refreshCovers();
      });
      // 3. The private scope itself needs to refresh its dependencies
      this._private.available.refresh();

      this.toDispose.push(this._private);
    }
    return this._private;
  }

  /**
   * Find a variable field by key path in the public scope by scope chain.
   * @param keyPath - The key path of the variable field.
   * @returns The variable field, or undefined if not found.
   */
  getByKeyPath(keyPath: string[]): BaseVariableField | undefined {
    return this.public.available.getByKeyPath(keyPath);
  }

  /**
   * Find a variable field by key path in the private scope by scope chain.
   * @param keyPath - The key path of the variable field.
   * @returns The variable field, or undefined if not found.
   */
  getByKeyPathInPrivate(keyPath: string[]): BaseVariableField | undefined {
    return this.private?.available.getByKeyPath(keyPath);
  }
}
