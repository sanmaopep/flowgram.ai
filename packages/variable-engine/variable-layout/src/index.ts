export { FlowNodeVariableData } from './flow-node-variable-data';
export { FreeLayoutScopeChain } from './chains/free-layout-scope-chain';
export { VariableLayoutConfig } from './variable-layout-config';
export { FixedLayoutScopeChain } from './chains/fixed-layout-scope-chain';
export {
  type FlowNodeScopeMeta,
  type FlowNodeScope,
  FlowNodeScopeTypeEnum as FlowNodeScopeType,
} from './types';
export { GlobalScope, bindGlobalScope } from './scopes/global-scope';
