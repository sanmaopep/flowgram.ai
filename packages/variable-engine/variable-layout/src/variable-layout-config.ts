import { Scope } from '@flowgram.ai/variable-core';
import { VariableEngine } from '@flowgram.ai/variable-core';
import { FlowNodeEntity, FlowDocument } from '@flowgram.ai/document';

import { type FlowNodeScope, type ScopeChainNode } from './types';

interface TransformerContext {
  scope: FlowNodeScope;
  document: FlowDocument;
  variableEngine: VariableEngine;
}

export interface VariableLayoutConfig {
  /**
   * 节点的子节点输出变量，不能被后续节点所访问，用于固定布局场景
   * @param node
   * @returns
   */
  isNodeChildrenPrivate?: (node: ScopeChainNode) => boolean;

  /**
   * 用于自由画布场景，部分场景通过连线或者其他交互形式来表达节点之间的父子关系，需要可配置化
   */
  getFreeChildren?: (node: FlowNodeEntity) => FlowNodeEntity[];
  getFreeParent?: (node: FlowNodeEntity) => FlowNodeEntity | undefined;

  /**
   * 对依赖作用域进行微调
   */
  transformDeps?: (scopes: Scope[], ctx: TransformerContext) => Scope[];

  /**
   * 对依赖作用域进行微调
   */
  transformCovers?: (scopes: Scope[], ctx: TransformerContext) => Scope[];
}

export const VariableLayoutConfig = Symbol('VariableLayoutConfig');
