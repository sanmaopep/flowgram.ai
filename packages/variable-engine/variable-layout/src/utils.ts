/**
 * Copyright (c) 2025 Bytedance Ltd. and/or its affiliates
 * SPDX-License-Identifier: MIT
 */

import { FlowNodeEntity } from '@flowgram.ai/document';

import { FlowNodeVariableData } from './flow-node-variable-data';

/**
 * Use `node.scope` instead
 * @deprecated
 * @param node
 */
export function getNodeScope(node: FlowNodeEntity) {
  return node.getData(FlowNodeVariableData).public;
}

/**
 * Use `node.privateScope` instead
 * @deprecated
 * @param node
 */
export function getNodePrivateScope(node: FlowNodeEntity) {
  return node.getData(FlowNodeVariableData).initPrivate();
}
