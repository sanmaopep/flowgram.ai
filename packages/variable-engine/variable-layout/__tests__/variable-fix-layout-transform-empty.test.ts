import { test, expect } from 'vitest';

import { ScopeChainTransformService } from '../src';
import { runFixedLayoutTest } from '../__mocks__/run-fixed-layout-test';
import { freeLayout1 } from '../__mocks__/free-layout-specs';

runFixedLayoutTest('Variable Fixed Layout transform empty', freeLayout1, {
  onInit(container) {
    const transformService = container.get(ScopeChainTransformService);

    transformService.registerTransformer('MOCK', {
      transformCovers: (scopes) => scopes,
      transformDeps: (scopes) => scopes,
    });

    // again transformer, prevent duplicated transformerId
    transformService.registerTransformer('MOCK', {
      transformCovers: () => [],
      transformDeps: () => [],
    });
    transformService.registerTransformer('MOCK', {
      transformCovers: () => [],
      transformDeps: () => [],
    });
  },
  runExtraTest: (container) => {
    test('check has transformer', () => {
      const transformService = container.get(ScopeChainTransformService);
      expect(transformService.hasTransformer('MOCK')).to.be.true;
      expect(transformService.hasTransformer('VARIABLE_LAYOUT_CONFIG')).to.be.false;
      expect(transformService.hasTransformer('NOT_EXIST')).to.be.false;
    });
  },
});
