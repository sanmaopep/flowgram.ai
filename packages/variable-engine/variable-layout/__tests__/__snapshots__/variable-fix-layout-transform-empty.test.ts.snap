// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Variable Fixed Layout transform empty > test get Covers 1`] = `
Map {
  "start_0" => [],
  "end_0" => [],
  "base_1" => [],
  "base_2" => [],
  "loop_1" => [],
  "base_in_loop_1" => [],
  "base_in_loop_2" => [],
  "base_in_loop_3" => [],
  "base_3" => [],
}
`;

exports[`Variable Fixed Layout transform empty > test get Covers After Init Private 1`] = `
Map {
  "start_0" => [],
  "end_0" => [],
  "base_1" => [],
  "base_2" => [],
  "loop_1" => [],
  "base_in_loop_1" => [],
  "base_in_loop_2" => [],
  "base_in_loop_3" => [],
  "base_3" => [],
}
`;

exports[`Variable Fixed Layout transform empty > test get Deps 1`] = `
Map {
  "start_0" => [],
  "end_0" => [],
  "base_1" => [],
  "base_2" => [],
  "loop_1" => [],
  "base_in_loop_1" => [],
  "base_in_loop_2" => [],
  "base_in_loop_3" => [],
  "base_3" => [],
}
`;

exports[`Variable Fixed Layout transform empty > test get Deps After Init Private 1`] = `
Map {
  "start_0" => [],
  "end_0" => [],
  "base_1" => [],
  "base_2" => [],
  "loop_1" => [],
  "base_in_loop_1" => [],
  "base_in_loop_2" => [],
  "base_in_loop_3" => [],
  "base_3" => [],
}
`;

exports[`Variable Fixed Layout transform empty > test get private scope Covers 1`] = `
Map {
  "start_0_private" => [],
  "end_0_private" => [],
  "base_1_private" => [],
  "base_2_private" => [],
  "loop_1_private" => [],
  "base_in_loop_1_private" => [],
  "base_in_loop_2_private" => [],
  "base_in_loop_3_private" => [],
  "base_3_private" => [],
}
`;

exports[`Variable Fixed Layout transform empty > test get private scope Deps 1`] = `
Map {
  "start_0_private" => [],
  "end_0_private" => [],
  "base_1_private" => [],
  "base_2_private" => [],
  "loop_1_private" => [],
  "base_in_loop_1_private" => [],
  "base_in_loop_2_private" => [],
  "base_in_loop_3_private" => [],
  "base_3_private" => [],
}
`;

exports[`Variable Fixed Layout transform empty > test sort 1`] = `
[
  "start_0",
  "testScope",
  "end_0",
  "base_1",
  "base_2",
  "loop_1",
  "base_in_loop_1",
  "base_in_loop_2",
  "base_in_loop_3",
  "base_3",
  "start_0_private",
  "end_0_private",
  "base_1_private",
  "base_2_private",
  "loop_1_private",
  "base_in_loop_1_private",
  "base_in_loop_2_private",
  "base_in_loop_3_private",
  "base_3_private",
]
`;
